src/log.rs

```diff
@@ -0,0 +1,106 @@
+//! Logging adapter for Laminar
+//!
+//! This module implements a simple, threaded-logger-friendly logging adapter. Logging adapters are
+//! used to attach an arbitrary logger into Laminar.
+use std::fmt;
+use std::sync::Arc;
+
+/// Logger trait for laminar
+///
+/// Any user of Laminar can implement this trait to attach their favorite logger to an instance of
+/// laminar. The log levels correspond to the same log levels as in the `log` crate.
+pub trait LaminarLogger {
+    /// Log a trace message
+    fn trace(&self, disp: Displayer);
+    /// Log a debug message
+    fn debug(&self, disp: Displayer);
+    /// Log an info message
+    fn info(&self, disp: Displayer);
+    /// Log a warning message
+    fn warn(&self, disp: Displayer);
+    /// Log an error message
+    fn error(&self, disp: Displayer);
+}
+
+// ---
+
+/// Holds a handle to a formatter function while implementing the [fmt::Display] trait.
+pub struct Displayer {
+    data: Arc<dyn Fn(&mut ::std::fmt::Formatter) -> ::std::fmt::Result + Send + Sync>,
+}
+
+impl Displayer {
+    pub(crate) fn new(
+        delegate: Arc<dyn Fn(&mut ::std::fmt::Formatter) -> ::std::fmt::Result + Send + Sync>,
+    ) -> Self {
+        Self { data: delegate }
+    }
+}
+
+impl fmt::Display for Displayer {
+    fn fmt(&self, f: &mut ::std::fmt::Formatter) -> ::std::fmt::Result {
+        (self.data)(f)
+    }
+}
+
+// ---
+
+pub(crate) struct DefaultLogger;
+
+impl LaminarLogger for DefaultLogger {
+    fn trace(&self, _: Displayer) {}
```

Comment:
I'm not sure. On one hand a library should be silent if not configured to log.

Reply:
Right, seems like a good point, just moving decisions over to the user. We might implement a logger using `log` and move it under the `tester` feature flag. That would allow us to debug if we provide our tester flag

Reply:
I honestly think that the default logger should maintain its current behavior but we can also provide a `SilentLogger` which is this implementation you can configure when setting up your connection.